Documentation/Plugin-developer-manual.md

```diff
@@ -0,0 +1,230 @@
+
+### Module API overview
+
+This section explains some general concepts encountered when implementing a module API. Most of this section will apply to most module types. For more detailed information see the module specific section, header files or the doxygen documentation.
+
+Modules with configuration data define an *INSTANCE* object, which is created by the module code in a `createInstance`-function or equivalent. The instance creation function is called during MaxScale startup, usually when creating services. The module instance data is held by the *SERVICE*-structure (or other higher level construct) and given as a parameter when calling functions from the module in question. The instance structure should contain all non-client-specific information required by the functions of the module. The core does not know what the object contains (since it is defined by the module itself), nor will it modify the pointer or the referenced object in any way. In the interface, the instance pointer is defined as `void*`, but in actual module code the pointer should always be cast to the actual instance type.
+
+Modules dealing with client-specific data require a *SESSION* object for every client. As with the instance data, the definition of the module session structure is up to the writer and MaxScale treats it as an opaque type. Usually the session contains status indicators and any resources required by the client. MaxScale core has its own `MXS_SESSION` object, which tracks a variety of client related information. The `MXS_SESSION` is given as a parameter to module-specific session creation functions and is required for several typical operations such as connecting to backends.
+
+Descriptor control blocks (`DCB`), are generalized I/O descriptor types. DCBs store the file descriptor, state, remote address, username, session, and other data. DCBs are created whenever a new socket is created. Typically this happens when a new client connects or MaxScale connects the client session to backend servers. The module writer should use DCB handling functions provided by the CPI to manage connections instead of calling general networking libraries.
+
+Network data such as client queries and backend replies are held in a rather complicated buffer container type called `GWBUF`. Multiple GWBUFs can form a linked list with type information and properties in each GWBUF-node. Each node includes a pointer to a reference counted shared buffer (`SHARED_BUF`), which finally points to a slice of the actual data. What all this means is that multiple GWBUF-chains can share some parts of data while having some parts private. The construction is meant to minimize the need for data copying. In simple situations there is no need to form a linked list and the buffer acts similar to a handle. Plugin writers should use the CPI to manipulate GWBUFs.
```

Comment:
Hmm, I'm not sure it is rather complicated as it's basically just a sharable linked list of buffers :)

If you append to a GWBUF you'll have a linked list. So, it's very easy to get one. Also, unless the filter/router module in its capabilities specify that it want contiguous buffers then the module writer must assume that it gets non-contiguous ones and act accordingly.